### DISCUSSION TITLE
Event-Stream Module Discussion

### DISCUSSION THREAD

**Discussion by benmusson** [mussonindustrial/embr#176] [General]
## Event-Stream Discussion
A place for general discussion related to the [**Event-Stream module**](https://github.com/mussonindustrial/embr/tree/main/modules/event-stream).

**Comment by benmusson** [comment]
@iandoyle-hbt
> https://github.com/mussonindustrial/testcontainers-ignition/discussions/18#discussioncomment-10455432
> Also, I've had success creating an Embr Event Stream with version 0.2.0. I'm really looking forward to SSE.

Care to share your use case for the Event-Stream module?
I've been mainly developing it as a way to supply real-time data to the **Charts** module, so I'm interested in your application.

**Reply by iandoyle-dev** [reply]
I was working with [Daniel Talbot's / Artek Integrated Solutions'](https://inductiveautomation.com/exchange/2463/overview) navigation component from the Ignition Exchange.

In this resource, they embed a WebDev endpoint into an iframe on the view. From this iframe, they reference the parentWindow.document to gain access to the computed styles. This allows them to keep track of the rendered state and toggle classes—really creative. I've only worked within the same document before, so referencing or inheriting from the parent seemed like an interesting concept to explore.

It's related to the issue that [Nader Chinichian](https://forum.inductiveautomation.com/t/perspective-mouse-click-drag/28583/53) illustrates here.

Using the iframe technique, I was able to bind a tag to an input, embed a WebDev resource in an iframe, reference the parentWindow element, and listen to onChange events, which provide data to a canvas element. Now, I have a heatmap canvas element with real-time data. It performs well, but the implementation feels like a short-term solution.

Server-sent events are more aligned with web standards and offer a more reliable way to consume data.

**Reply by benmusson** [reply]
Oh wow, very interesting.

I don't know if you've seen, but version 0.4.0 refactored things into [ `streams`](https://github.com/mussonindustrial/embr/tree/main/modules/event-stream/gateway/src/main/kotlin/com/mussonindustrial/embr/eventstream/streams). By subscribing to a stream, you'll receive all events that occur on the stream over the SSE endpoint.

 Currently there is `TagStream` and `LicenseStream`; the `TagStream` sends real time tag information, tag alarms, and a way to query/stream history, and the `LicenseStream` sends license status changes.

An idea I've been mulling around is a `PerspectiveMessageStream` that retransmits Perspective session-scoped messages (you would have to authenticate using a Perspective session id). I think it would be a pretty convenient way to communicate with an iFrame, it'd be one directional but pretty low effort.
